File: 2021/cpp/day_13.cpp
//
// Created by trox667 on 13.12.21.
//

#include <iostream>
#include <set>
#include "day_13.hpp"
#include "utils.hpp"


namespace day13 {

    using Point = std::pair<int, int>;
    using Instruction = std::pair<char, int>;

    void get_coords_instructions(const std::vector<std::string> &lines,
                                 std::vector<Point> &coords,
                                 std::vector<Instruction> &instructions) {
        for (const auto &line: lines) {
            if (line.starts_with("fold")) {
                const auto tokens = utils::split(line, '=');
                if (tokens[0].ends_with('y')) {
                    instructions.emplace_back(
                            std::make_pair('y', std::stoi(tokens[1])));
                } else {
                    instructions.emplace_back(
                            std::make_pair('x', std::stoi(tokens[1])));
                }
            } else if (!line.empty()) {
                const auto tokens = utils::split(line, ',');
                coords.emplace_back(std::stoi(tokens[0]), std::stoi(tokens[1]));
            }
        }
    }

    void print_coords(const std::vector<Point> &coords) {
        auto maxx = 0;
        auto maxy = 0;
        for (const auto &coord: coords) {
            if (std::get<0>(coord) > maxx) maxx = std::get<0>(coord);
            if (std::get<1>(coord) > maxy) maxy = std::get<1>(coord);
        }
        for (auto y = 0; y <= maxy; ++y) {
            for (auto x = 0; x <= maxx; ++x) {
                if (std::find_if(coords.begin(), coords.end(),
                                 [x, y](const auto &coord) {
                                     return std::get<0>(coord) == x &&
                                            std::get<1>(coord) == y;
                                 }) != coords.end()) {
                    std::cout << '#';
                } else {
                    std::cout << '.';
                }
            }
            std::cout << std::endl;
        }
    }

    [[nodiscard]] std::vector<Point>
    fold(const Instruction &instruction, const std::vector<Point> &coords) {
        auto result = std::set<Point>{};
        for (const auto &coord: coords) {
            const auto x = std::get<0>(coord);
            const auto y = std::get<1>(coord);
            const auto value = std::get<1>(instruction);
            if (std::get<0>(instruction) == 'y') {
                auto ty = value - std::abs(y - value);
                result.insert(std::make_pair(x, ty));
            } else {
                auto tx = value - std::abs(x - value);
                result.insert(std::make_pair(tx, y));
            }
        }
        return std::vector<Point>(result.begin(), result.end());
    }


    [[nodiscard]] long part1(const std::vector<std::string> &lines) {
        auto coords = std::vector<Point>{};
        auto instructions = std::vector<Instruction>{};
        get_coords_instructions(lines, coords, instructions);

        for (const auto &instruction: instructions) {
            coords = fold(instruction, coords);
            break;
        }

        return coords.size();
    }

    [[nodiscard]] long part2(const std::vector<std::string> &lines) {
        auto coords = std::vector<Point>{};
        auto instructions = std::vector<Instruction>{};
        get_coords_instructions(lines, coords, instructions);

        for (const auto &instruction: instructions) {
            coords = fold(instruction, coords);
        }
        std::cout << std::endl;
        print_coords(coords);

        return 0;
    }
}

File: 2021/rust/src/day13.rs
use std::collections::HashSet;

type Instruction = (char, isize);
type Point = (isize, isize);
fn create_instructions_coords(lines: &Vec<String>) -> (Vec<Instruction>, Vec<Point>) {
    let mut instructions = vec![];
    let mut coords = vec![];
    for line in lines {
        if line.is_empty() {
            continue;
        } else if line.starts_with("fold") {
            let direction: Vec<&str> = line.split('=').collect();
            if direction[0].ends_with("y") {
                instructions.push(('y', direction[1].parse::<isize>().unwrap()));
            } else {
                instructions.push(('x', direction[1].parse::<isize>().unwrap()));
            }
        } else {
            let xy: Vec<&str> = line.split(',').collect();
            coords.push((
                xy[0].parse::<isize>().unwrap(),
                xy[1].parse::<isize>().unwrap(),
            ));
        }
    }
    (instructions, coords)
}

fn print_coords(coords: &Vec<Point>) {
    let maxx = coords.iter().max_by_key(|coord| coord.0).unwrap();
    let maxy = coords.iter().max_by_key(|coord| coord.1).unwrap();
    for y in 0..=maxy.1 {
        for x in 0..=maxx.0 {
            if coords.contains(&(x, y)) {
                print!("#");
            } else {
                print!(".");
            }
        }
        println!();
    }
}

fn fold(instruction: &Instruction, coords: &Vec<Point>) -> Vec<Point> {
    let mut result = HashSet::<Point>::new();
    for (x, y) in coords {
        let value = instruction.1;
        if instruction.0 == 'y' {
            result.insert((*x, value - (y - value).abs()));
        } else {
            result.insert((value - (x - value).abs(), *y));
        }
    }
    Vec::from_iter(result.into_iter())
}

pub fn part1(lines: &Vec<String>) -> usize {
    let (instructions, mut coords) = create_instructions_coords(lines);
    for instruction in instructions {
        coords = fold(&instruction, &coords);
        break;
    }
    coords.len()
}

pub fn part2(lines: &Vec<String>) -> usize {
    let (instructions, mut coords) = create_instructions_coords(lines);
    for instruction in instructions {
        coords = fold(&instruction, &coords);
    }
    print_coords(&coords);
    0
}
